From f501c955e86eb3043ab5e3c5f2648ed4c8bb58a3 Mon Sep 17 00:00:00 2001
From: Hannes Reinecke <hare@suse.de>
Date: Thu, 11 Dec 2014 15:46:22 +0100
Subject: [PATCH] 95fcoe: Store current configuration in dracut cmdline

When running with --hostonly-cmdline we should be storing
the current configuration in /etc/cmdline.d so that dracut
will be configure the system automatically.

References: bnc#877288

Signed-off-by: Hannes Reinecke <hare@suse.de>
Signed-off-by: Thomas Renninger <trenn@suse.de>
(cherry picked from commit 8b257fcc86f9a470613c1a5719641144a688acf7)
---
 modules.d/95fcoe/module-setup.sh | 45 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 45 insertions(+)

diff --git a/modules.d/95fcoe/module-setup.sh b/modules.d/95fcoe/module-setup.sh
index c000b3be..af2798dc 100755
--- a/modules.d/95fcoe/module-setup.sh
+++ b/modules.d/95fcoe/module-setup.sh
@@ -13,6 +13,7 @@ check() {
     }
 
     require_binaries dcbtool fipvlan lldpad ip readlink fcoemon fcoeadm || return 1
+
     return 0
 }
 
@@ -25,6 +26,46 @@ installkernel() {
     instmods fcoe 8021q edd
 }
 
+get_vlan_parent() {
+    local link=$1
+
+    [ -d $link ] || return
+    read iflink < $link/iflink
+    for if in /sys/class/net/* ; do
+	read idx < $if/ifindex
+	if [ $idx -eq $iflink ] ; then
+	    echo ${if##*/}
+	fi
+    done
+}
+
+# called by dracut
+cmdline() {
+
+    for c in /sys/bus/fcoe/devices/ctlr_* ; do
+        [ -L $c ] || continue
+        read enabled < $c/enabled
+        [ $enabled -eq 0 ] && continue
+        d=$(cd -P $c; echo $PWD)
+        i=${d%/*}
+        read mac < ${i}/address
+        s=$(dcbtool gc ${i##*/} dcb | sed -n 's/^DCB State:\t*\(.*\)/\1/p')
+        if [ -z "$s" ] ; then
+	    p=$(get_vlan_parent ${i})
+	    if [ "$p" ] ; then
+	        s=$(dcbtool gc ${p} dcb | sed -n 's/^DCB State:\t*\(.*\)/\1/p')
+	    fi
+        fi
+        if [ "$s" = "on" ] ; then
+	    dcb="dcb"
+        else
+	    dcb="nodcb"
+        fi
+        echo "fcoe=${mac}:${dcb}"
+    done
+}
+
+# called by dracut
 install() {
     inst_multiple ip dcbtool fipvlan lldpad readlink lldptool fcoemon fcoeadm
     inst_libdir_file 'libhbalinux.so*'
@@ -33,6 +74,10 @@ install() {
     mkdir -m 0755 -p "$initdir/var/lib/lldpad"
     mkdir -m 0755 -p "$initdir/etc/fcoe"
 
+    if [[ $hostonly_cmdline == "yes" ]] ; then
+        local _fcoeconf=$(cmdline)
+        [[ $_fcoeconf ]] && printf "%s\n" "$_fcoeconf" >> "${initdir}/etc/cmdline.d/95fcoe.conf"
+    fi
     inst "$moddir/fcoe-up.sh" "/sbin/fcoe-up"
     inst "$moddir/fcoe-edd.sh" "/sbin/fcoe-edd"
     inst "$moddir/fcoe-genrules.sh" "/sbin/fcoe-genrules.sh"
